Fields must fit in the type, even for repr(Rust) (#2166)
rust-lang/reference: src/type-layout.md

r[layout.repr.rust.layout]
The only data layout guarantees made by this representation are those required for soundness. These are:

 1. The offset of a constructible field is divisible by that field's alignment.
 2. The alignment of the type is at least the maximum alignment of its constructible fields.
 3. For any constructible field, its offset plus its size is at most the size of the type.

r[layout.repr.rust.layout.constructible]

> A field is considered constructible if it is possible to create a value of the type containing the field.

Author: I think under this section of the reference, those xs *have* to be separate fields because otherwise a field might have multiple offsets, and this section talks about fields as if they have a single offset.

The only data layout guarantees made by this representation are those required for soundness. They are:

 1. The fields are properly aligned.

>  2. The fields do not overlap and fit in the struct.

Reviewer: The surrounding text is about " All user-defined composite types (structs, enums, and unions) " (`layout.repr.intro`). If some part of this text is meant to apply only to structs, that needs to be made clear upfront, not in the middle of a section.

Reviewer: > For structs, the second guarantee means that

This is very strange -- re-interpreting what "fits in" means depending on the kind of type? I would rather have two separate guarantees:
1. All fields fit in the type, i.e., the offset of each field plus its size does not exceed the size of the type.
2. All fields in a struct, and all fields within a single enum variant, do not overlap. (For zero-sized fields, this implies that the field is not contained within the range of another field.)

Either way this would make a guarantee about enums that has been quite contentious during the offset_of_enum discussion (and ultimately prevented stabilizing that feature) so I am not sure we're willing to commit to this.

 1. The offset of a field is divisible by that field's alignment.
 2. The alignment of the type is at least the maximum alignment of its fields.
 3. For any field, its offset plus its size is at most the size of the type.

For enums, the above guarantees only apply to fields of inhabited variants.

r[layout.repr.rust.layout.struct]

> For [structs], it is further guaranteed that the fields do not overlap. That is, the fields can be ordered such that the offset plus the size of any field is less than or equal to the offset of the next field in the ordering, or for the last field, the size of the struct. The ordering does not have to be the same as the order in which the fields are specified in the declaration of the type.

Reviewer: ```suggestion
For [structs], it is further guaranteed that the fields do not overlap. That is, the fields are ordered such that the offset plus the size of any field is less than or equal to the offset of the next field in the ordering, or for the last field, the size of the struct. The ordering does not have to be the same as the order in which the fields are specified in the declaration of the type.
```
They actually are ordered, right? "can be ordered" is confusing to me.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2166/91f144460fcb93bed2b580887aa59c28d662a10b..b1f9d0596f54ca58d0e683a09cb6cd5a49335daa)*

Author: Due to ZST the ordering may not be unique.

Reviewer: Yeah, that's true. I still think "are ordered" is more clear than "can be".

"can be" sounds like one *could* do this but the actual layout algorithm doesn't actually do this.

Reviewer: ```suggestion
For [structs], it is further guaranteed that the fields do not overlap. That is, if we take the fields ordered by their offset, then for each field its offset plus its size is less than or equal to the offset of the next field in the ordering, or for the last field, the size of the struct. The ordering does not have to be the same as the order in which the fields are specified in the declaration of the type.
```
